src/librustc_typeck/check/writeback.rs

```diff
@@ -159,8 +160,56 @@ impl<'cx, 'gcx, 'tcx> WritebackCx<'cx, 'gcx, 'tcx> {
             _ => {}
         }
     }
+
+    // Similar to operators, indexing is always assumed to be overloaded
+    // Here, correct cases where an indexing expression can be simplified
+    // to use builtin indexing because the index type is known to be
+    // usize-ish
+    fn fix_index_builtin_expr(&mut self, e: &hir::Expr) {
+        if let hir::ExprIndex(ref base, ref index) = e.node {
+            let mut tables = self.fcx.tables.borrow_mut();
+
+            let base_ty = {
+                let base_ty = tables.expr_ty_adjusted(&base);
+                // When unsizing, the final type of the expression is taken
+                // from the first argument of the indexing operator, which
+                // is a &self, and has to be deconstructed
+                if let ty::TyRef(_, ref ref_to) = base_ty.sty {
+                    ref_to.ty
+                } else {
+                    base_ty
+                }
```

Comment:
This makes me a bit ... nervous. It seems strange to strip out the `&T` based purely on the type involved. That is, if we were looking at the *adjustment* and deciding to strip off a layer of references, that'd be one thing, but I deciding based on the type alone makes me nervous. Can you give some examples of why this line of code was needed?

Reply:
I'd guess `TyError`, but we should probably skip doing this at all if there are any errors.

Reply:
The Unsize adjustment made by the index typechecking code (`2252-2257` in this codebase) doesn't have the unsized type as the `target`, instead taking `method.sig.inputs()[0]`, a `&self`. This is, to my understanding, the only way for the result of `expr_ty_adjusted` to result in a `TyRef`, since otherwise a reference would be deref'd, and you can't do an `impl<'a, T> Index<usize> for &'a [T]` of some kind.

`builtin_index()` returns `None` on `TyRef`, regardless of the underlying type - I considered changing this instead, but I felt this would confuse the dereferencing logic. Alternatively, I could introduce some kind of `fn unsize_type` to use in place of the trait argument, which would return `TySlice` for `TyArray`, which is the only case I can think of that matters.

Reply:
Actually, on a second closer look, this just requires using `self_ty`, which *is* the correct unsized type, and not `method.sigs.inputs()[0]`, in that adjustment.

Reply:
Unsizing coercions happen on references and raw/smart pointers, so you first have a borrow to get `&[T; N]` and then an unsizing of that reference to `&[T]`.

Reply:
@ivanbakel my point was that I would feel more comfortbale "skipping" the ref if we only did it when the coercion was an `Unsize` coercion -- i.e., I don't like deciding to skip based on the *type*, that feels like it could misfire and skip when we didn't expect.

Reply:
What @eddyb said made me re-examine the behaviour, and this ref skip is always necessary, because it's always introduced in the indexing logic even without an `Unsize`. I've rewritten the code to reflect that. Let me know if you still have the same concern.